File: kernel/src/main/kotlin/ninja/blacknet/network/packet/Pong.kt
/*
 * Copyright (c) 2018-2024 Pavel Vasin
 *
 * Licensed under the Jelurida Public License version 1.1
 * for the Blacknet Public Blockchain Platform (the "License");
 * you may not use this file except in compliance with the License.
 * See the LICENSE.txt file at the top-level directory of this distribution.
 */

package ninja.blacknet.network.packet

import kotlinx.serialization.Serializable
import ninja.blacknet.network.Connection

@Serializable
class Pong(
    private val response: Int,
) : Packet {
    override fun handle(connection: Connection) {
        val (challenge, requestTime) = connection.pingRequest ?: return connection.dos("Unexpected packet Pong")

        val solution = if (connection.version >= Ping.MIN_VERSION)
            solve(challenge)
        else if (connection.version == 13)
            solveV1(challenge)
        else
            challenge

        if (response != solution) {
            connection.dos("Invalid Pong")
            return
        }

        connection.ping = connection.lastPacketTime - requestTime
        connection.pingRequest = null
    }
}

File: network/src/main/rust/packet/pong.rs
/*
 * Copyright (c) 2018-2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use crate::connection::Connection;
use crate::packet::{Packet, Ping, PingV1};
use serde::{Deserialize, Serialize};
use std::sync::Arc;

#[derive(Deserialize, Serialize)]
pub struct Pong {
    response: u32,
}

impl Pong {
    pub const fn new(response: u32) -> Self {
        Self { response }
    }
}

impl Packet for Pong {
    fn handle(self, connection: &Arc<Connection>) {
        if let Some((challenge, request_time)) = connection.ping_request() {
            let magic = connection.node().mode().network_magic();
            let solution = if connection.version() >= Ping::MIN_VERSION {
                Ping::solve(magic, challenge)
            } else if connection.version() == 13 {
                PingV1::solve(magic, challenge)
            } else {
                challenge
            };

            if self.response != solution {
                connection.dos("Invalid Pong");
                return;
            }

            connection.set_ping(connection.last_packet_time() - request_time);
            connection.set_ping_request(None);
        } else {
            connection.dos("Unexpected packet Pong");
        }
    }
}
